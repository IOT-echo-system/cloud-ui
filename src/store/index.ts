import type {TRootActions, TRootReducer, TRootState} from '@/typing/store'
import * as siteActions from './actions/site'
import * as userActions from './actions/user'
import siteReducer, {initSiteState} from './reducers/site'
import userReducer, {initUserState} from '@/store/reducers/user'

const combineReducers = <S = TRootState>(reducers: {[K in keyof S]: TRootReducer<S[K]>}): TRootReducer<S> => {
  return (state: S, action: TRootActions): S => {
    return (Object.keys(reducers) as Array<keyof S>).reduce(
      (prevState: S, key: keyof S) => ({...prevState, [key]: reducers[key](prevState[key], action)}),
      state
    )
  }
}

export const rootState = {
  site: initSiteState,
  user: initUserState
}

export const rootActions = {
  site: siteActions,
  user: userActions
}

export const rootReducer = combineReducers({
  site: siteReducer,
  user: userReducer
})
