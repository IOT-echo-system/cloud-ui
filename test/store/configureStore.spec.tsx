import type {Dispatch} from 'react'
import React, {useContext} from 'react'
import {act, render} from '@testing-library/react'
import StoreProvider, {GlobalContext} from '@/store/configureStore'
import type {TRootActions, TRootState} from '@/typing/store'
import {rootState} from '@/store'
import {siteBuilder} from '../builders/stateBuilder'

describe('Configure Store Test', () => {
  beforeEach(jest.resetAllMocks)
  afterEach(jest.resetAllMocks)

  it('should match with the snapshot', () => {
    const {container} = render(
      <StoreProvider>
        <div />
      </StoreProvider>
    )

    expect(container).toMatchSnapshot()
  })

  it('should give the initial State', () => {
    let store: {state: TRootState; dispatch: Dispatch<TRootActions>}

    function TestComponent() {
      store = useContext(GlobalContext)!
      return null
    }

    const {container} = render(
      <StoreProvider>
        <TestComponent />
      </StoreProvider>
    )

    expect(container).toMatchSnapshot()
    expect(store!.state).toStrictEqual(rootState)
  })

  it('should update State on dispatch call', () => {
    let store: {state: TRootState; dispatch: Dispatch<TRootActions>}

    function TestComponent() {
      store = useContext(GlobalContext)!
      return null
    }

    const {container} = render(
      <StoreProvider>
        <TestComponent />
      </StoreProvider>
    )

    expect(container).toMatchSnapshot()
    expect(store!.state).toStrictEqual(rootState)

    expect(store!.state.site.title).toStrictEqual('Cloud')
    const site = siteBuilder({title: 'updated Title'})
    act(() => {
      store.dispatch({type: 'SITE_UPDATE_STATE', payload: {site}})
    })
    expect(store!.state.site.title).toStrictEqual('updated Title')
  })
})
