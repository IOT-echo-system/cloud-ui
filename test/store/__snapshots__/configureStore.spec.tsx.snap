// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configure Store Test should give the initial State 1`] = `<div />`;

exports[`Configure Store Test should match with the snapshot 1`] = `
<div>
  <div />
</div>
`;

exports[`Configure Store Test should update State on dispatch call 1`] = `<div />`;
